### PR TITLE
Merge go1.5+ assembly

### PR DESCRIPTION
**Pull request by tamird**
See commit message.

### CODE REVIEW

**Review by nvanbenschoten — COMMENTED** [review]
Implementing the assembly fast path for i386 seems like a great change. I don't understand why it needs to be grouped in with the rest of this refactor, especially if the refactor is not a clear win because it introduces a new function call on amd64.

What's your primary goal here?

**Comment by tamird** [comment]
The goal was just to reduce the mental overhead of reading this code. Regarding the added function call: I see no difference on local benchmarks:

Before:
```
$ go test -v -bench .
=== RUN   TestGet
--- PASS: TestGet (0.00s)
goos: linux
goarch: amd64
pkg: github.com/petermattis/goid
cpu: AMD Ryzen 9 7950X 16-Core Processor            
BenchmarkGet
BenchmarkGet-32         1000000000               0.9726 ns/op
PASS
ok      github.com/petermattis/goid     1.073s
```

After:
```
$ go test -v -bench .
=== RUN   TestGet
--- PASS: TestGet (0.00s)
goos: linux
goarch: amd64
pkg: github.com/petermattis/goid
cpu: AMD Ryzen 9 7950X 16-Core Processor            
BenchmarkGet
BenchmarkGet-32         1000000000               0.9563 ns/op
PASS
ok      github.com/petermattis/goid     1.076s
```

I suspect that's because `Get` itself is always inlined, so the only real function call is into `getg`.

**Review by nvanbenschoten — APPROVED** [review]
LGTM after discussing on Slack and determining that `Get` can be inlined. From @tamird:
```
$ go build -gcflags='-m' 2>&1 | grep -F 1.5
./goid_go1.5.go:26:6: can inline Get
```

**Review by nvanbenschoten — APPROVED** [review]
LGTM after discussing on Slack and determining that `Get` can be inlined. From @tamird:
```
$ go build -gcflags='-m' 2>&1 | grep -F 1.5
./goid_go1.5.go:26:6: can inline Get
```

**Comment by tamird** [comment]
Added CI coverage for inlining as well.

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -128,18 +131,47 @@ jobs:
       - name: 'Build with ${{ matrix.go }}'
         if: ${{ matrix.arch == 'x64' }}
         run: go build -v ./...
+      - name: 'Build with ${{ matrix.go }}'
+        if: ${{ matrix.arch == '386' }}
+        env:
+          GOARCH: 386
+        run: go build -v ./...
+      - name: 'Check that Get is inlined with ${{ matrix.go }}'
+        if: ${{ matrix.arch == 'x64' }}
```
Comment (nvanbenschoten) [inline .github/workflows/go.yml:140]: > if echo -e '${{ matrix.go }}\n1.12' | sort -V | head -n1 | xargs test 1.12 = ; then

Maybe you'll find this just as hacky, but what about:
```
# NOTE: matrix.go comparison uses lexicographical ordering.
if: ${{ matrix.arch == 'x64' && matrix.go < '1.5' && matrix.go >= '1.12' }}
```

Comment (tamird) [reply .github/workflows/go.yml:140]: That'll silently do the wrong thing when go1.50 comes out :laughing:

**Pull request merged**